java/src/main/native/src/jni_writer_data_sink.hpp

```diff
@@ -0,0 +1,177 @@
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy =
+          left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char *copy_to = current_buffer_data + current_buffer_written;
+
+      std::memcpy(copy_to, copy_from, amount_to_copy);
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  bool supports_device_write() const override { return true; }
+
+  void device_write(void const *gpu_data, size_t size, rmm::cuda_stream_view stream) override {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    long left_to_copy = static_cast<long>(size);
+    const char *copy_from = static_cast<const char *>(gpu_data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        stream.synchronize();
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy =
+          left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char *copy_to = current_buffer_data + current_buffer_written;
+
+      CUDF_CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost,
+                                    stream.value()));
+
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+    stream.synchronize();
+  }
+
+  std::future<void> device_write_async(void const *gpu_data, size_t size,
+                                       rmm::cuda_stream_view stream) override {
+    // Call the sync version until figuring out how to write asynchronously.
+    device_write(gpu_data, size, stream);
+    return std::async(std::launch::deferred, [] {});
+  }
+
+  void flush() override {
+    if (current_buffer_written > 0) {
+      JNIEnv *env = cudf::jni::get_jni_env(jvm);
+      handle_buffer(env, current_buffer, current_buffer_written);
+      if (current_buffer != nullptr) {
+        env->DeleteGlobalRef(current_buffer);
+      }
+      current_buffer = nullptr;
+      current_buffer_len = 0;
+      current_buffer_data = nullptr;
+      current_buffer_written = 0;
+    }
+  }
+
+  size_t bytes_written() override { return total_written; }
+
+  void set_alloc_size(long size) { this->alloc_size = size; }
+
+private:
+  void rotate_buffer(JNIEnv *env) {
```

Comment:
Not a huge fan of this name. It implies a double-buffering scenario to me.

Reply:
Agreed, there. It took me a while to grok the intention of that code.